cluster/v1alpha1/types.go

```diff
@@ -0,0 +1,101 @@
+// A workload can be defined to deployed on a ManagedClusterSet, which mean:
+//
+// 1. The workload can run on any ManagedCluster in the ManagedClusterSet
+// 2. The workload cannot run on any ManagedCluster outside the ManagedClusterSet
+// 3. The service exposed by the workload can be shared in any ManagedCluster in the ManagedClusterSet
+type ManagedClusterSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the attributes of the desired ManagedClusters
+	Spec ManagedClusterSetSpec `json:"spec"`
+
+	// Status represents the current status of the ManagedClusterSet
+	// +optional
+	Status ManagedClusterSetStatus `json:"status,omitempty"`
+}
+
+// ManagedClusterSetSpec describes the attributes of the desired ManagedClusters
+type ManagedClusterSetSpec struct {
+	// ClusterSelectors represents a slice of selectors to select ManagedClusters
+	// If empty, the ManagedClusterSet will include all ManagedClusters
+	// If more than one ClusterSelector are specified in the slice, OR operation
+	// will be used between them.
+	// +optional
+	ClusterSelectors []ClusterSelector `json:"clusterSelectors,omitempty"`
+}
+
+// ClusterSelector represents a selector of ManagedClusters
+// ClusterNames and LabelSelector are mutually exclusive. They cannot be set at the
+// same time. If none of them is set, all ManagedClusters will be selected
+type ClusterSelector struct {
+	// ClusterNames represents a list of cluster name
+	// +optional
+	ClusterNames []string `json:"clusterNames,omitempty"`
+
+	// LabelSelector represents a label selector to select cluster by label
+	// +optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+}
+
+// ManagedClusterSetStatus represents the current status of the ManagedClusterSet.
+type ManagedClusterSetStatus struct {
+	// Conditions contains the different condition statuses for this ManagedClusterSet.
+	Conditions []StatusCondition `json:"conditions"`
+}
+
+// StatusCondition contains condition information for a ManagedClusterSet.
+type StatusCondition struct {
+	// Type is the type of the ManagedClusterSet condition.
+	// +required
+	Type string `json:"type"`
+
+	// Status is the status of the condition. One of True, False, Unknown.
+	// +required
+	Status metav1.ConditionStatus `json:"status"`
+
+	// LastTransitionTime is the last time the condition changed from one status to another.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+
+	// Reason is a (brief) reason for the condition's last status change.
+	// +required
+	Reason string `json:"reason"`
+
+	// Message is a human-readable message indicating details about the last status change.
+	// +required
+	Message string `json:"message"`
+}
+
+const (
+	// ManagedClusterSetConditionEmpty means no ManagedCluster is included in the
+	// ManagedClusterSet.
+	ManagedClusterSetConditionEmpty string = "ClusterSetEmpty"
```

Comment:
Like paul, I wonder about the utility of this vs representing how many it selected as an integer and condition saying you matched.  But I can go with it for now.